### PR TITLE
Implementation of EnsembleModel

### PR DESCRIPTION
**Pull request by anachal**
Vincent will probably tell me that it's spelled "EnsembleModel"

### CODE REVIEW

**Comment by anachal** [comment]
Is it good?

File: notebooks/EnsembleModel.ipynb [hunk 1]
```diff
@@ -0,0 +1,207 @@
+{
```
Comment (koaning) [inline notebooks/EnsembleModel.ipynb]: please remove this notebook for now.

File: sklego/ensemble_model.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import confusion_matrix
+
+class EnsembleModel():
+    """
+    A classifier that takes the uncertainty p(class | model) of a classifier into account.
+    Based on the requirements of issue #23 of koaning/scikit-lego
+    """
+
+    def __init__(self, X_test, y_test, estimator=LogisticRegression(), alpha=0.5):
```
Comment (koaning) [inline sklego/ensemble_model.py]: So ... 

... I think it might be a good idea to have a seperate validation set on which to base the confusion matrix. However this object needs to primarily fit a `sklearn.pipeline` so the base behavior should be to base the confusion matrix on the training set. 

I also think it might be nicer not to use `LogisticRegression()` as a base input and instead demand that the user defines this explicitly. Explicit > Implicit.

Comment (koaning) [reply sklego/ensemble_model.py]: Also this might be a better order. 

```
def __init__(self, estimator=LogisticRegression(), alpha=0.5, X_test=None, y_test=None)
```

File: sklego/ensemble_model.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        self.y_test = y_test
+
+    def fit(self, X, y):
+        """
+
+        :param X: array of feature values used for training the classifier
+        :param y: array of target values used for training the classifier
+        :return: trained classifier (estimator)
+        """
+        self.fitted_estimator_ = self.estimator.fit(X, y)
```
Comment (koaning) [inline sklego/ensemble_model.py]: perhaps a different name than `self.fitted_estimator_` maybe ... `self.orig_estimator` 

also you may want to calculate the confusion matrix in the `self.fit` as well

File: sklego/ensemble_model.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        """
+
+        :param X: array of feature values upon which a prediction is to be made
+        :return: Updated prediction probabilities of each of the target value made by the trained classifier
+        """
+        p = self.predict_proba(X)
+        prob_matrix = self.prob_matrix()
+        return p @ prob_matrix
+
+    def make_prediction(self, X):
```
Comment (koaning) [inline sklego/ensemble_model.py]: this needs to be called `.predict(X)` ... better yet ... implement just `predict_proba` and you get `.predict()` for free.

File: tests/test_estimators/test_ensemble_model.py [hunk 5]
```diff
@@ -0,0 +1,33 @@
+from sklego.ensemble_model import EnsembleModel
+from sklearn import datasets
+from sklearn.model_selection import train_test_split
+
+# Data
```
Comment (koaning) [inline tests/test_estimators/test_ensemble_model.py]: this needs to be put into a fixture.

File: tests/test_estimators/test_ensemble_model.py [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+iris = datasets.load_iris()
+X = iris.data[:, :2]  # we only take the first two features.
+y = iris.target
+X_train, X_test, y_train, y_test = train_test_split(X, y)
+
+# Model
+model = EnsembleModel(X_test = X_test, y_test = y_test)
+model.fit(X_train, y_train)
+
+def test_alpha():
```
Comment (koaning) [inline tests/test_estimators/test_ensemble_model.py]: better: test that if `alpha > 1` an error gets thrown.

File: tests/test_estimators/test_ensemble_model.py [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+    assert model.alpha >= 0
+    assert model.alpha <= 1
+
+
+def test_predict_proba():
```
Comment (koaning) [inline tests/test_estimators/test_ensemble_model.py]: in this case there is a module in the testing framework where shapes and other common things are tested for **all** estimators. might be better to put this test there.

File: sklego/ensemble_model.py (hunk 3 above)
Comment (MBrouns) [inline sklego/ensemble_model.py]: `sklearn` expects that the parameters (`self.estimator` in this case) don't change in `fit` so it's probably best to clone the self.estimator (see for example the `EstimatorTransformer`)

```
 self.estimator_ = clone(self.estimator)
 self.estimator_.fit(X, y)
```

File: sklego/ensemble_model.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        :param X: array of feature values upon which a prediction is to be made
+        :return: prediction of the target value made by the trained classfifier
+        """
+        return self.fitted_estimator_.predict(X)
+
+    def predict_proba(self, X):
+        """
+
+        :param X: array of feature values upon which a prediction is to be made
```
Comment (MBrouns) [inline sklego/ensemble_model.py]: You probably want to check if the estimator is fitted. You can use `check_is_fitted` from `sklearn.utils.validation` like this `check_is_fitted(self, 'estimator_')`

File: sklego/ensemble_model.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
+from sklearn.linear_model import LogisticRegression
+from sklearn.metrics import confusion_matrix
+
+class EnsembleModel():
+    """
+    A classifier that takes the uncertainty p(class | model) of a classifier into account.
+    Based on the requirements of issue #23 of koaning/scikit-lego
```
Comment (MBrouns) [inline sklego/ensemble_model.py]: It might be nice to refer to the original paper here

Comment (koaning) [reply sklego/ensemble_model.py]: It's not *that* great of a paper. But sure.

**Comment by koaning** [comment]
Actually ... come to think of it ... this model is not an emsemble model as much. It's more of a postprocessor. @MBrouns what might be a better submodule? `postprocessor` or `meta`?

**Comment by MBrouns** [comment]
Agreed that it's not an ensemble, more like a `CMReweigher`. To me it feels more like a `meta` than a postprocessor as it actually wraps the estimator

**Comment by koaning** [comment]
This package is bound to have a lot of models in `meta`. Do you forsee any way to maybe split it up?

**Comment by koaning** [comment]
Referencing the issue: https://github.com/koaning/scikit-lego/issues/23

File: sklego/ensemble_model.py (hunk 4 above)
Comment (anachal) [reply sklego/ensemble_model.py]: For some reason I couldn't manage to get `.predict()` for free, so I defined it. How can I get it like that?

File: sklego/ensemble_model.py (hunk 9 above)
Comment (anachal) [reply sklego/ensemble_model.py]: Where can I find the paper?

File: sklego/ensemble_model.py (hunk 4 above)
Comment (MBrouns) [reply sklego/ensemble_model.py]: I think you only get `predict` if you mixin `LinearClassifierMixin` but we probably don't want that in this case so implementing it yourself is fine

File: sklego/ensemble_model.py (hunk 9 above)
Comment (MBrouns) [reply sklego/ensemble_model.py]: https://link.springer.com/chapter/10.1007/978-3-319-00651-2_19

File: sklego/ensemble_model.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+       :param y: array of target values used for training the classifier
+       :return:
+                - trained classifier (self.orig_estimator)
+                - confusion matrix populated with the probabilities of each real value vs model prediction value
+                  of the target variable (self.confusion_matrix)
+                - updated prediction probabilities of each target value made by the trained classifier
+                  (self.prob_matrix)
+       """
+        self.estimator_ = clone(self.estimator)
+        self.orig_estimator_ = self.estimator_.fit(X, y)
```
Comment (MBrouns) [inline sklego/ensemble_model.py]: You now have `self.estimator_` and `self.orig_estimator_` referring to the same estimator, I would get rid of one of them

**Comment by koaning** [comment]
In terms of a name convention. `EnsembleModel` might be a bit too vague and the added feature here is the addition of `ReweightedEnsemble`?
